pkg/apis/core/validation/validation.go

```diff
@@ -3542,9 +3542,8 @@ func validateInitContainers(containers []core.Container, regularContainers []cor
 				allErrs = append(allErrs, field.Forbidden(idxPath.Child("startupProbe"), "may not be set for init containers without restartPolicy=Always"))
 			}
 		}
-
-		if len(ctr.ResizePolicy) > 0 {
-			allErrs = append(allErrs, field.Invalid(idxPath.Child("resizePolicy"), ctr.ResizePolicy, "must not be set for init containers"))
+		if !isRestartableInitContainer(&ctr) && len(ctr.ResizePolicy) > 0 {
```

Comment:
Yes, I think it's possible to get into a state of failing validation on an existing pod on rollback:

1. On v1.32, create a pod with a container & sidecar that both set a ResizePolicy
2. Roll back to v1.31
3. ResizePolicy fields [don't get dropped](https://github.com/kubernetes/kubernetes/blob/release-1.31/pkg/api/pod/util.go#L657): `InPlacePodVerticalScaling=false`, but [`inPlacePodVerticalScalingInUse`](https://github.com/kubernetes/kubernetes/blob/release-1.31/pkg/api/pod/util.go#L1057) still returns true
4. Fails validation: https://github.com/kubernetes/kubernetes/blob/release-1.31/pkg/apis/core/validation/validation.go#L3541

Reply:
These are our options I can think of:

1. Move forward with this PR, but guard most of the changes with a new alpha feature gate
2. Relax validation with the a new alpha feature gate, but postpone the bulk of the implementation to next release
3. Treat a nil resize policy as an implicit default (maybe we should do this anyway), and relax validation with a new alpha feature gate.

Option 3 means we can still add sidecar resize support in v1.32 beta, but if you want to force a container restart on resize you need to wait until v1.33.

WDYT?

Reply:
Option 1 is the "normal" path?

I don't think I understand option 2.

Option 3: I am not sure I follow - treat nil as RestartNotRequired?  Would you disallow explicit values, to prevent rollback problems?